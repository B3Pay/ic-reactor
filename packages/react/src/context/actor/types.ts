import type {
  IDL,
  ActorHooksReturnType,
  BaseActor,
  ActorManagerParameters,
  CanisterId,
  InitializeActor,
} from "../../types"

export interface CreateActorContextType<A = BaseActor>
  extends ActorHooksReturnType<A> {
  useInitializeActor?: () => InitializeActor
}

export interface CreateActorContextReturnType<A = BaseActor>
  extends ActorHooksReturnType<A> {
  ActorProvider: React.FC<ActorProviderProps>
  ActorHookProvider: React.FC<ActorHookProviderProps<A>>
  useInitializeActor: () => InitializeActor
}

export interface ActorHookProviderProps<A> {
  hooks: ActorHooksReturnType<A>
  children?: React.ReactNode
}

export interface ActorProviderProps extends CreateActorContextParameters {
  children?: React.ReactNode | undefined
  candidString?: string
  fetchingComponent?: React.ReactNode
  errorComponent?: (error: string) => React.ReactNode
  loadingComponent?: React.ReactNode
  authenticatingComponent?: React.ReactNode
}

export interface CreateActorContextParameters
  extends Omit<
    ActorManagerParameters,
    "idlFactory" | "agentManager" | "canisterId"
  > {
  didjsId?: string
  canisterId?: CanisterId
  idlFactory?: IDL.InterfaceFactory
  disableAutoFetch?: boolean
}

export interface ActorChildrenProps extends React.PropsWithChildren {
  useActorState: ActorHooksReturnType["useActorState"]
}
