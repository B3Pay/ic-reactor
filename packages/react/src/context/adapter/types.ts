import type { PropsWithChildren } from "react"
import type { CandidAdapterParameters, IDL } from "../../types"
import type { CandidAdapter } from "@ic-reactor/core/dist/classes"

export type { CandidAdapter }

export interface CandidAdapterContextType extends CandidAdapter {}

export interface CandidAdapterProviderProps
  extends PropsWithChildren,
    CandidAdapterParameters {
  withParser?: boolean
  loadingComponent?: React.ReactNode
}

export interface CreateCandidAdapterCotextParameters
  extends CandidAdapterParameters {
  withParser?: boolean
}

export interface UseCandidEvaluationReturnType {
  fetchError: string | null
  fetching: boolean
  evaluateCandid: () => Promise<IDL.InterfaceFactory | undefined>
}

export interface CreateCandidAdapterContextReturnType {
  CandidAdapterContext: React.Context<CandidAdapterContextType | null>
  CandidAdapterProvider: React.FC<CandidAdapterProviderProps>
  useCandidEvaluation: (candidString: string) => UseCandidEvaluationReturnType
  useCandidAdapter: () => CandidAdapter
}
