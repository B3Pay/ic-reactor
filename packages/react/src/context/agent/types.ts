import type { PropsWithChildren } from "react"
import type {
  AgentHooksReturnType,
  AuthHooksReturnType,
  AgentManager,
  AgentManagerParameters,
} from "../../types"

export interface AgentContext
  extends AgentHooksReturnType,
    AuthHooksReturnType {
  agentManager: AgentManager
}

export interface CreateAgentCotextParameters extends AgentManagerParameters {
  withProcessEnv?: boolean
  disableAuthenticateOnMount?: boolean
}

export interface CreateAgentContextReturnType
  extends AgentHooksReturnType,
    AuthHooksReturnType {
  useAgentManager: () => AgentManager
  AgentProvider: React.FC<AgentProviderProps>
  AgentContext: React.Context<AgentContext | null>
}

export interface AgentProviderProps
  extends PropsWithChildren,
    CreateAgentCotextParameters {
  agentManager?: AgentManager
}
