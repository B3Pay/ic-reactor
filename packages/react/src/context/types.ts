import type {
  IDL,
  ActorHooksReturnType,
  AgentHooksReturnType,
  AuthHooksReturnType,
  BaseActor,
  AgentManager,
  ActorManagerParameters,
  AgentManagerParameters,
  CanisterId,
} from "../types"
import type { PropsWithChildren } from "react"

export interface AgentContext
  extends AgentHooksReturnType,
    AuthHooksReturnType {
  agentManager: AgentManager
}

export interface CreateAgentCotextParameters extends AgentManagerParameters {
  withProcessEnv?: boolean
  disableAuthenticateOnMount?: boolean
}

export interface CreateAgentContextReturnType
  extends AgentHooksReturnType,
    AuthHooksReturnType {
  useAgentManager: () => AgentManager
  AgentProvider: React.FC<AgentProviderProps>
}

export interface AgentProviderProps
  extends PropsWithChildren,
    CreateAgentCotextParameters {
  agentManager?: AgentManager
}

export interface CreateActorContextReturnType<A = BaseActor>
  extends ActorHooksReturnType<A> {
  ActorProvider: React.FC<ActorProviderProps>
}

export interface ActorProviderProps extends CreateActorContextParameters {
  children?: React.ReactNode | undefined
  candidString?: string
  loadingComponent?: React.ReactNode
  authenticatingComponent?: React.ReactNode
}

export interface CreateActorContextParameters
  extends Omit<
    ActorManagerParameters,
    "idlFactory" | "agentManager" | "canisterId"
  > {
  didjsId?: string
  canisterId?: CanisterId
  idlFactory?: IDL.InterfaceFactory
  initializeParser?: boolean
}
