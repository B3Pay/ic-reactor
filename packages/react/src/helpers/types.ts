import type { AuthClientLoginOptions } from "@dfinity/auth-client"
import type { getAgentHooks } from "./agent"
import type { getAuthHooks } from "./auth"
import type {
  ActorState,
  CanisterId,
  ActorMethodArgs,
  ActorMethodReturnType,
  Identity,
  Principal,
  FunctionName,
  VisitService,
  AuthClient,
} from "@ic-reactor/core/dist/types"

export type AgentHooks = ReturnType<typeof getAgentHooks>

export type AuthHooks = ReturnType<typeof getAuthHooks>

export interface UseAuthClientArgs {
  onAuthentication?: (promise: () => Promise<Identity>) => void
  onAuthenticationSuccess?: (identity: Identity) => void
  onAuthenticationFailure?: (error: Error) => void
  onLoginSuccess?: (principal: Principal) => void
  onLoginError?: (error: Error) => void
  onLogin?: (promise: () => Promise<Principal>) => void
  onLoggedOut?: () => void
}

export interface UseAuthClientReturn {
  error: Error | undefined
  authClient: AuthClient | null
  authenticated: boolean
  authenticating: boolean
  identity: Identity | null
  login: (options?: LoginOptions) => Promise<void>
  logout: (options?: LogoutOptions) => Promise<void>
  authenticate: () => Promise<Identity>
  loginLoading: boolean
  loginError: Error | null
}

export type LoginState = {
  loading: boolean
  error: Error | null
}

export type LoginOptions = AuthClientLoginOptions

export type LogoutOptions = { returnTo?: string }

export type ReactorCallArgs<A, M extends FunctionName<A>> = {
  functionName: M
  args?: ActorMethodArgs<A[M]>
  onLoading?: (loading: boolean) => void
  onError?: (error: Error | undefined) => void
  onSuccess?: (data: ActorMethodReturnType<A[M]> | undefined) => void
  throwOnError?: boolean
}

export type ActorCallState<A, M extends FunctionName<A>> = {
  data: ActorMethodReturnType<A[M]> | undefined
  error: Error | undefined
  loading: boolean
}

export interface UseQueryCallArgs<A, M extends FunctionName<A>>
  extends ReactorCallArgs<A, M> {
  refetchOnMount?: boolean
  refetchInterval?: number | false
}

export interface UseUpdateCallArgs<A, M extends FunctionName<A>>
  extends ReactorCallArgs<A, M> {}

export interface ReactorCallReturn<
  A,
  M extends FunctionName<A> = FunctionName<A>
> extends ActorCallState<A, M> {
  reset: () => void
  call: (
    eventOrReplaceArgs?: React.MouseEvent | ActorMethodArgs<A[M]>
  ) => Promise<ActorMethodReturnType<A[M]> | undefined>
}

export type ReactorCall<A> = <M extends FunctionName<A>>(
  args: ReactorCallArgs<A, M>
) => ReactorCallReturn<A, M>

export type UseQueryCall<A> = <M extends FunctionName<A>>(
  args: UseQueryCallArgs<A, M>
) => ReactorCallReturn<A, M>

export type UseUpdateCall<A> = <M extends FunctionName<A>>(
  args: UseUpdateCallArgs<A, M>
) => ReactorCallReturn<A, M>

export type UseMethodCall<A> = <M extends FunctionName<A>>(
  args: UseMethodCallArg<A, M>
) => UseMethodCallReturn<A, M>

export interface UseMethodCallReturn<
  A,
  M extends FunctionName<A> = FunctionName<A>
> extends ReactorCallReturn<A, M> {
  visit: VisitService<A>[M]
  reset: () => void
  error: Error | undefined
  loading: boolean
}

export type UseMethodCallArg<A, M extends FunctionName<A>> = ReactorCallArgs<
  A,
  M
>

export interface UseActorState extends Omit<ActorState, "methodState"> {
  canisterId: CanisterId
}

export interface ActorHooks<A> {
  initialize: () => Promise<void>
  useActorState: () => UseActorState
  useQueryCall: UseQueryCall<A>
  useUpdateCall: UseUpdateCall<A>
  useMethodCall: UseMethodCall<A>
  useVisitMethod: <M extends FunctionName<A>>(
    functionName: M
  ) => VisitService<A>[M]
}
