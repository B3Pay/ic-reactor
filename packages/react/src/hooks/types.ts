import {
  IDL,
  ActorHooksReturnType,
  ActorManagerParameters,
  BaseActor,
  CanisterId,
} from "../types"

export type ActorReConfigParameters = Omit<
  ActorManagerParameters,
  "idlFactory" | "canisterId" | "agentManager"
>

export interface UseActorParameters extends ActorReConfigParameters {
  candidString?: string
  canisterId: CanisterId
  idlFactory?: IDL.InterfaceFactory
  disableAutoFetch?: boolean
}

export interface UseActorReturn<A = BaseActor> {
  hooks: ActorHooksReturnType<A> | null
  fetching: boolean
  fetchError: string | null
  authenticating: boolean
  initializeActor: InitializeActor
}

export type InitializeActor = (
  idlFactory: IDL.InterfaceFactory,
  actorReConfig?: ActorReConfigParameters
) => void
