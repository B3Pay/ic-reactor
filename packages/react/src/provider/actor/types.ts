import { IDL } from "@dfinity/candid"
import { ActorManagerOptions, BaseActor } from "@ic-reactor/core/dist/types"
import { ActorHooks, AgentContextType } from "../../types"

export interface ActorContextType<Actor = BaseActor> extends ActorHooks<Actor> {
  ActorContext: React.Context<ActorContextType<Actor> | null>
  useActorContext: <A = Actor>() => ActorContextType<A>
  ActorProvider: React.FC<ActorProviderProps>
}

export type CreateReactorContext = <A = BaseActor>(
  options?: Partial<CreateActorOptions>
) => ActorContextType<A>

export interface CreateActorOptions
  extends Omit<
    ActorManagerOptions,
    "idlFactory" | "agentManager" | "canisterId"
  > {
  didjsId?: string
  canisterId?: string
  agentContext?: AgentContextType
  idlFactory?: IDL.InterfaceFactory
  loadingComponent?: React.ReactNode
}

export interface ActorProviderProps extends CreateActorOptions {
  children?: React.ReactNode | undefined
  loadingComponent?: React.ReactNode
}
