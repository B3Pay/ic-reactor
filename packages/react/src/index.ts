import type { AuthClientLoginOptions } from "@dfinity/auth-client"
import type { ActorSubclass, ReActorOptions } from "@ic-reactor/store"
import { createReActorStore } from "@ic-reactor/store"
import { useEffect, useState } from "react"
import { useStore } from "zustand"
import { getCallHooks } from "./hooks"

export * from "./context"

export const createReActor = <A extends ActorSubclass<any>>(
  options: ReActorOptions
) => {
  const isLocal =
    typeof process !== "undefined" &&
    (process.env.NODE_ENV === "development" ||
      process.env.DFX_NETWORK === "local")

  const { callMethod, initialize, authenticate, authStore, actorStore } =
    createReActorStore<A>({
      isLocal,
      ...options,
    })

  const useActorStore = () => {
    const actorState = useStore(actorStore, (state) => state)

    return { ...actorState, initialize }
  }

  const useAuthStore = () => {
    const authState = useStore(authStore, (state) => state)
    return authState
  }

  const useAuthClient = () => {
    const [loginLoading, setLoginLoading] = useState(false)
    const [loginError, setLoginError] = useState<Error | unknown | null>(null)
    const { authClient, authenticated, authenticating, identity } =
      useAuthStore()

    const login = async (options?: AuthClientLoginOptions) => {
      setLoginLoading(true)
      setLoginError(null)

      try {
        await authClient?.login({
          ...options,
          onSuccess: async () => {
            setLoginLoading(false)
            await authenticate()
            options?.onSuccess?.()
          },
          onError: (e) => {
            setLoginError(e)
            setLoginLoading(false)
            options?.onError?.(e)
          },
        })
      } catch (e) {
        setLoginLoading(false)
        setLoginError(e)
      }
    }

    const logout = async (options?: { returnTo?: string }) => {
      await authClient?.logout(options)
      await authenticate()
    }

    useEffect(() => {
      if (!authClient && !authenticating) {
        authenticate()
      }
    }, [authClient, authenticating])

    return {
      authClient,
      authenticated,
      authenticating,
      identity,
      login,
      logout,
      authenticate,
      loginLoading,
      loginError,
    }
  }

  const { useQueryCall, useMethodField, useMethodFields, useUpdateCall } =
    getCallHooks<A>(callMethod, actorStore)

  return {
    useMethodFields,
    useMethodField,
    useActorStore,
    useAuthStore,
    useQueryCall,
    useUpdateCall,
    useAuthClient,
  }
}
