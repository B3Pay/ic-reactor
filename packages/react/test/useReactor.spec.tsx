import React from "react"
import renderer, { act } from "react-test-renderer"
import { backend, idlFactory } from "./candid"
import { useReactor } from "../src"
import { AgentProvider } from "../src"
import { ActorHooks } from "../src/types"

describe("createReactor", () => {
  it("should query", async () => {
    const BackendActor = () => {
      const { hooks, fetchError, fetching } = useReactor<typeof backend>({
        canisterId: "xeka7-ryaaa-aaaal-qb57a-cai",
        idlFactory,
      })

      return fetching ? (
        <p>Loading Candid interface...</p>
      ) : fetchError ? (
        <p>Error: {fetchError}</p>
      ) : (
        hooks && <TestComponent {...hooks} />
      )
    }

    const TestComponent = ({ useQueryCall }: ActorHooks<typeof backend>) => {
      const { data: version, loading } = useQueryCall({
        functionName: "version",
      })

      return (
        <div>
          <span>
            {loading
              ? "Loading..."
              : version
              ? version.toString()
              : "Ready To call"}
          </span>
        </div>
      )
    }

    let screen = renderer.create(
      <AgentProvider>
        <BackendActor />
      </AgentProvider>
    )

    const versionStatus = () => screen.root.findAllByType("span")[0]

    expect(screen.toJSON()).toMatchSnapshot()

    expect(versionStatus().props.children).toEqual("Ready To call")

    await act(async () => {
      await new Promise((r) => setTimeout(r, 100))
    })

    expect(screen.toJSON()).toMatchSnapshot()

    expect(versionStatus().props.children).toEqual("Loading...")

    await act(async () => {
      await new Promise((r) => setTimeout(r, 1000))
    })

    expect(versionStatus().props.children).toEqual("0.2.0")

    expect(screen.toJSON()).toMatchSnapshot()
  })
})
