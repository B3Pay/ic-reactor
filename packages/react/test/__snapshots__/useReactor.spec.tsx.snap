// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createReactor should query 1`] = `
<div>
  <span>
    Ready To call
  </span>
</div>
`;

exports[`createReactor should query 2`] = `
<div>
  <span>
    Loading...
  </span>
</div>
`;

exports[`createReactor should query 3`] = `
<div>
  <span>
    0.2.0
  </span>
</div>
`;
